### PR TITLE
V2 final

### PR DESCRIPTION
**Pull request by billytrend-cohere**
This PR introduces a new version of the Cohere API, v2, which includes improvements to the Chat, Classify, Embed, and Rerank APIs. It also introduces a new SDK, `cohere.ClientV2`.

## Changes
- The `model` parameter is now required for Embed, Rerank, Classify, and Chat.
- The `embedding_types` parameter is now required for Embed.
- The `messages` parameter now consists of a list of roles (`system`, `user`, `assistant`, or `tool`). The `system` role in v2 replaces the `preamble` parameter in v1.
- Chat history is now managed via the `messages` list.
- Response content is now accessed via `message.content[0].text`.
- The `seed` and `temperature` parameters of the Chat API can be used to control the predictability of the model's output.
- The `system` role in the `messages` parameter can be used to provide instructions for the model.
- The `documents` parameter now supports a few different options for structuring documents: a list of objects with `data` object, a list of objects with `data` string, or a list of strings.
- The `message.citations` parameter is now used to access citations.
- Cited documents are now accessed as part of `message.citations`, in the `sources` field.
- The `search_queries_only` parameter is now supported via tools.
- The `connectors` parameter is now supported via user-defined tools.
- The `web-search` connector in the `connectors` parameter is now supported via tools.
- The `conversation_id` parameter is no longer supported (chat history is now managed by the developer via the `messages` parameter).
- The `search_queries_only`, `connectors`, and `prompt_truncation` parameters are no longer supported.
- The `force_single_step` parameter is no longer supported (all tool calls are now multi-step by default).
- The `tool_call_id` parameter is now used to emit tool call IDs.
- Tool execution is now handled by appending `tool_call_id` and `tool_content` to `messages` to the chat history.
- User message is now set as empty (`""`) by default.
- The `citations` parameter is now used to access citations.
- Cited tools are now accessed as part of `message.citations`, in the `sources` field.
- The `citation_quality` parameter is now controlled via the `citation_options` parameter (with `mode` as a key).
- The following v1 features are not supported in v2: general chat, RAG, tool use, and unsupported features in v2.

### CODE REVIEW

File: fern/pages/v2/text-generation/chat-api.mdx [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  token: '<YOUR API KEY>',
+});
+
+(async () => {
+  const response = await cohere.chat({
+    message: 'Write a title for a blog post about API design. Only output the title text.',
+  });
+
+  console.log(response.text)
+})();
```
Comment (mrmer1) [inline fern/pages/v2/text-generation/chat-api.mdx:63]:
```suggestion
public class ChatPost {
  public static void main(String[] args) {
    Cohere cohere = Cohere.builder().token("<YOUR API KEY>").build();

    NonStreamedChatResponse response = cohere.chat(
      ChatRequest.builder()
      	.model("command-r-plus-08-2024")
      	.message("Write a title for a blog post about API design. Only output the title text.")
    )

    System.out.println(response); // "The Art of API Design: Crafting Elegant and Powerful Interfaces"
  }
}
```

File: fern/pages/v2/text-generation/chat-api.mdx [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    NonStreamedChatResponse response = cohere.chat(
+      ChatRequest.builder()
+      	.model("command-r-plus-08-2024")
+      	.message("Write a title for a blog post about API design. Only output the title text.")
+    )
+
+    System.out.println(response); // "The Art of API Design: Crafting Elegant and Powerful Interfaces"
+  }
+}
```
Comment (mrmer1) [inline fern/pages/v2/text-generation/chat-api.mdx:48]:
```suggestion
package chatv2post;

import com.cohere.api.Cohere;
import com.cohere.api.resources.v2.requests.V2ChatRequest;
import com.cohere.api.types.*;
import java.util.List;

public class Default {
    public static void main(String[] args) {
        Cohere cohere = Cohere.builder().token("<<apiKey>>").clientName("snippet").build();

        ChatResponse response =
                cohere.v2()
                        .chat(
                                V2ChatRequest.builder()
                                    .model("command-r-plus")
                                    .messages(
                                        List.of(
                                            ChatMessageV2.user(
                                                UserMessage.builder()
                                                    .content(
                                                        UserMessageContent
                                                                .of("Hello world!"))
                                                    .build())))
                                    .build());

        System.out.println(response);
    }
}

```

File: fern/pages/v2/text-generation/chat-api.mdx (hunk 1 above)
Comment (mrmer1) [inline fern/pages/v2/text-generation/chat-api.mdx:63]:
```suggestion
const { CohereClientV2 } = require('cohere-ai');

const cohere = new CohereClientV2({
  token: '<<apiKey>>',
});

(async () => {
  const response = await cohere.chat({
    model: 'command-r-plus',
    messages: [
      {
        role: 'user',
        content: 'hello world!',
      },
    ],
  });

  console.log(response);
})();

```

File: fern/pages/v2/text-generation/retrieval-augmented-generation-rag.mdx [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+For example, you can customize the preamble to encourage a longer list of search queries to be generated.
+
+```python PYTHON
+instructions_verbose = "Write many search queries that will find helpful information for answering the user's question accurately. Always write a very long list of at least 7 search queries. If you decide that a search is very unlikely to find information that would be useful in constructing a response to the user, you should instead directly answer.
+```
+```
+# Sample response
+['NSync popularity', 'Backstreet Boys popularity', 'NSync vs Backstreet Boys popularity comparison', 'Which boy band is more popular NSync or Backstreet Boys', 'NSync and Backstreet Boys fan base size comparison', 'Who has sold more albums NSync or Backstreet Boys', 'NSync and Backstreet Boys chart performance comparison']
+```
```
Comment (mrmer1) [inline fern/pages/v2/text-generation/retrieval-augmented-generation-rag.mdx:205]:
```suggestion

```

Comment (mrmer1) [reply fern/pages/v2/text-generation/retrieval-augmented-generation-rag.mdx:205]: removing duplicate text

File: fern/pages/v2/text-generation/retrieval-augmented-generation-rag.mdx [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+#### Step 2: Fetching relevant documents
+
+The next step is to [fetch documents](https://docs.cohere.com/docs/documents-and-citations) from the relevant data source using the generated search queries. For example, to answer the question about the two pop sensations _NSYNC_ and _Backstreet Boys_, one might want to use an API from a web search engine, and fetch the contents of the websites listed at the top of the search results.
```
Comment (mrmer1) [inline fern/pages/v2/text-generation/retrieval-augmented-generation-rag.mdx:209]:
```suggestion
The next step is to fetch documents from the relevant data source using the generated search queries. For example, to answer the question about the two pop sensations _NSYNC_ and _Backstreet Boys_, one might want to use an API from a web search engine, and fetch the contents of the websites listed at the top of the search results.
```

File: fern/pages/v2/text-generation/tools/multi-step-tool-use.mdx [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+#### What is the difference between tool use and Retrieval Augmented Generation (RAG)?
+
+Tool use is a natural extension of retrieval augmented generation (RAG). RAG is about enabling the model to interact with an information retrieval system (like a vector database). Our models are trained to be excellent at RAG use cases.
+
+Tool use pushes this further, allowing Cohere models to go far beyond information retrieval, interact with search engines, APIs, functions, databases, and many other tools.
+
+## A Further Example With Multiple Tools
+
+This section provides another example of multi-step tool use, this time with multiple tools. The notebook for this example can be [found here](https://github.com/cohere-ai/notebooks/blob/main/notebooks/guides/agents/Multi_Step_Tool_Use_Spotify_v2.ipynb).
```
Comment (mrmer1) [inline fern/pages/v2/text-generation/tools/multi-step-tool-use.mdx:164]:
```suggestion
This section provides another example of multi-step tool use, this time with multiple tools. The notebook for this example can be [found here](https://github.com/cohere-ai/notebooks/blob/main/notebooks/agents/Multi_Step_Tool_Use_Spotify_v2.ipynb).
```

File: fern/pages/v2/tutorials/build-things-with-cohere/semantic-search-with-cohere.mdx [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+return_results(query_emb, doc_emb, faqs_long)
+```
+```
+Rank: 1
+Score: 0.442758615743984
+Document: {'text': "Finding Coffee Spots: For your caffeine fix, head to the break room's coffee machine or cross the street to the café for artisan coffee."}
+
+Rank: 2
+Score: 0.32783563708365726
+Document: {'text': 'Team-Building Activities: We foster team spirit with monthly outings and weekly game nights. Feel free to suggest new activity ideas anytime!'}
```
Comment (mrmer1) [inline fern/pages/v2/tutorials/build-things-with-cohere/semantic-search-with-cohere.mdx:136]:
```suggestion
Rank: 1
Score: 0.3872984617627964
Document: {'text': 'Team-Building Activities: We foster team spirit with monthly outings and weekly game nights. Feel free to suggest new activity ideas anytime!'}

Rank: 2
Score: 0.3272549670724577
Document: {'text': 'Proposing New Ideas: Innovation is welcomed! Share your brilliant ideas at our weekly team meetings or directly with your team lead.'}
```

File: fern/pages/v2/tutorials/build-things-with-cohere/rag-with-cohere.mdx [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+
+Let's now look at the first step—search query generation. The chatbot needs to generate an optimal set of search queries to use for retrieval. 
+
+There are different possible approaches to this. In this example, we'll take a [tool use](./v2/docs/tool-use) approach.
+
+Here, we build a tool that takes a user query and returns a list of relevant document snippets for that query. The tool can generate zero, one or multiple search queries depending on the user query.
+
+We'll not do a detailed walkthrough of tool use here as we'll do that in the next tutorial, which is a dedicated tutorial on tool use.
+
+We recommend using the v1 API for this functionality in order to leverage the `force_single_step` feature. Note: support for search query generation in v2 will be coming soon.
```
Comment (mrmer1) [inline fern/pages/v2/tutorials/build-things-with-cohere/rag-with-cohere.mdx:130]:
```suggestion
Here, we build a tool that takes a user query and returns a list of relevant document snippets for that query. The tool can generate zero, one or multiple search queries depending on the user query.
```

File: fern/pages/v2/tutorials/build-things-with-cohere/rag-with-cohere.mdx [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+        model="command-r-08-2024",
+        message=message,
+        force_single_step=True,
+        tools=query_gen_tool
+    )
+
+    if response.tool_calls:
+        search_queries = response.tool_calls[0].parameters["queries"]
+
+    return search_queries
```
Comment (mrmer1) [inline fern/pages/v2/tutorials/build-things-with-cohere/rag-with-cohere.mdx:168]:
```suggestion
def generate_search_queries(message: str) -> List[str]:
    
    # Define the query generation tool
    query_gen_tool = [
        {
            "type": "function",
            "function": {
                "name": "internet_search",
                "description": "Returns a list of relevant document snippets for a textual query retrieved from the internet",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "queries": {
                            "type": "array",
                            "items": {"type": "string"},
                            "description": "a list of queries to search the internet with.",
                        }
                    },
                    "required": ["queries"],
                },
            },
        }
    ]


    # Define a preamble to optimize search query generation
    instructions = "Write a search query that will find helpful information for answering the user's question accurately. If you need more than one search query, write a list of search queries. If you decide that a search is very unlikely to find information that would be useful in constructing a response to the user, you should instead directly answer."

    # Generate search queries (if any)
    search_queries = []
    
    res = co.chat(
        model="command-r-08-2024",
        messages=[
            {"role": "system", "content": instructions},
            {"role": "user", "content": message},
        ],
        tools=query_gen_tool
    )
    
    if res.message.tool_calls:
        for tc in res.message.tool_calls:
            queries = json.loads(tc.function.arguments)["queries"]
            search_queries.extend(queries)

    return search_queries
```

**Review by mkozakov — APPROVED** [review]

**Review by lfayoux — APPROVED** [review]

**Pull request merged**